csharp/src/Google.Protobuf/ExtensionSet.cs

```diff
@@ -0,0 +1,341 @@
+
+using Google.Protobuf.Collections;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Protobuf
+{
+    /// <summary>
+    /// Methods for managing <see cref="ExtensionSet{TTarget}"/>s with null checking.
+    /// 
+    /// Most users will not use this class directly
+    /// </summary>
+    public static class ExtensionSet
+    {
+        private static bool GetValue<TTarget>(ref ExtensionSet<TTarget> set, Extension extension, out IExtensionValue value) where TTarget : IExtendableMessage<TTarget>
+        {
+            if (set == null)
+            {
+                value = null;
+                return false;
+            }
+            return set.ValuesByNumber.TryGetValue(extension.FieldNumber, out value);
+        }
+
+        /// <summary>
+        /// Gets the value of the specified extension
+        /// </summary>
+        public static TValue Get<TTarget, TValue>(ref ExtensionSet<TTarget> set, Extension<TTarget, TValue> extension) where TTarget : IExtendableMessage<TTarget>
+        {
+            IExtensionValue value;
+            if (GetValue(ref set, extension, out value))
+            {
+                return ((ExtensionValue<TValue>)value).GetValue();
+            }
+            else 
+            {
+                return extension.DefaultValue;
+            }
+        }
+
+        /// <summary>
+        /// Gets the value of the specified repeated extension or null if it doesn't exist in this set
+        /// </summary>
+        public static RepeatedField<TValue> Get<TTarget, TValue>(ref ExtensionSet<TTarget> set, RepeatedExtension<TTarget, TValue> extension) where TTarget : IExtendableMessage<TTarget>
+        {
+            IExtensionValue value;
+            if (GetValue(ref set, extension, out value))
+            {
+                return ((RepeatedExtensionValue<TValue>)value).GetValue();
+            }
+            else 
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Gets the value of the specified repeated extension, registering it if it doesn't exist
+        /// </summary>
+        public static RepeatedField<TValue> GetOrRegister<TTarget, TValue>(ref ExtensionSet<TTarget> set, RepeatedExtension<TTarget, TValue> extension) where TTarget : IExtendableMessage<TTarget>
+        {
+            IExtensionValue value;
+            if (set == null)
+            {
+                value = extension.CreateValue();
+                set = new ExtensionSet<TTarget>();
+                set.ValuesByNumber.Add(extension.FieldNumber, value);
+            }
+            else
+            {
+                if (!set.ValuesByNumber.TryGetValue(extension.FieldNumber, out value))
+                {
+                    value = extension.CreateValue();
+                    set.ValuesByNumber.Add(extension.FieldNumber, value);
+                }
+            }
+
+            return ((RepeatedExtensionValue<TValue>)value).GetValue();

```

Comment:
needing to do this cast doesn't seem right.  The ExtensionValue classes are internal, so you can have methods that will allow you to avoid dirty tricks like this:
E.g. you can have two virtual methods/properties:  `GetSingleValue` and `GetRepeatedValue`  where GetSingleValue just throws InvalidOperationException for repeated extension and vice versa. The cast you're doing is basically a dirty way of achieving the same effect (and you also need to cast in multiple places in ExtensionSet instead of defining the behavior just once).

Reply:
I don't think I could do that with `IExtensionValue` since it's not generic. `GetSingleValue` and `GetRepeatedValue` on `IExtensionValue` would have to return an object instead of an actual generic value.

Reply:
Ok, got it. I think it would still be cleaner to have e.g. a `AsRepeatedExtension<TValue>()` method on IExtensionValue which would make the cast with the option to perform some extra checking. But after looking at the code, I cannot think of a failure scenario where the extension value would be of a wrong type (because we already know it's an extension targeting TTarget and we already checked the field number), so I think it's fine to leave the logic as is.
Thanks for the explanation.